Explore: Support user timezone

marefr opened the pull request:
**What this PR does / why we need it**:
Explore now uses the timezone of the user to decide if local browser time (current) or UTC should be used.

Some tests and cleanup probably needed.

**Which issue(s) this PR fixes**:
Closes #12812 

**Special notes for your reviewer**:
Would be nice with some initial feedback to know if I'm going in the right direction.

After looking into this a bit feels like timezone needs to be part of the url when going from dashboard to Explore and in that case it overrides the user timezone setting - what do you think?

marefr committed:
explore: support user timezone

hugohaggmark reviewed (approved):
LGTM!

As you said yourself, we should add this to the URL as well. But in a separate PR.

steven-sheehy commented:
This might be the wrong place, but is there any thought given to merging the "Timestamp" and "Local Time" columns in Log explorer? I've always found it confusing that there's multiple time columns. In my opinion there should be one that defaults to user time zone but maybe with an option to change it to UTC or other timezone.

marefr commented on public/app/features/explore/Graph.test.tsx:
```diff
@@ -8,6 +8,7 @@ const setup = (propOverrides?: object) => {
     size: { width: 10, height: 20 },
     data: mockData().slice(0, 19),
     range: { from: 'now-6h', to: 'now' },
+    timeZone: { raw: 'browser', isUtc: false },
```
Would be okay to add a constant `DefaultTimeZone` to grafana/ui?

marefr replied:
So for example a constant and a factory method in grafana/ui/types:

```typescript
export const parseTimeZone = (raw: string) => {
  return {
    raw: raw,
    isUtc: raw === 'utc',
  } as TimeZone;
};

export const DefaultTimeZone = parseTimeZone('browser');
```

marefr replied:
@torkelo @hugohaggmark any thoughts?

hugohaggmark replied:
```ts
export const parseTimeZone = (raw: string): TimeZone => {
  return {
    raw: raw,
    isUtc: raw === 'utc',
  };
};

export const DefaultTimeZone = parseTimeZone('browser');
```
Maybe move TimeZone to function declaration and skip the `as TimeZone part`.

marefr replied:
Nice!

marefr commented on packages/grafana-ui/src/types/time.ts:
```diff
@@ -15,3 +15,8 @@ export interface IntervalValues {
   intervalMs: number;
 }
+
+export interface Timezone {
```
Good point regarding naming. Fixed. 

> we need to think about how to support more than local + utc

Agree, but have same challenge in dashboard so not sure we're going to solve this in this PR. But agree we need to start think about it to be able to smoothly introduce more timezones without dependent plugins breaks and just works. Either we add date parsing utils to grafana/ui that takes the raw timezone or either we add `TimeZone` as a class with a function to convert/create `moment` or either a separate function to convert moment to specific timezone.

marefr commented:
@steven-sheehy yes I think your comment deserves a new issue. But good idea now when we're adding support for local browser time/utc.

marefr commented on public/app/features/explore/Graph.tsx:
```diff
@@ -134,23 +135,24 @@ export class Graph extends PureComponent<GraphProps, GraphState> {
     let { from, to } = range;
     if (!moment.isMoment(from)) {
-      from = dateMath.parse(from, false);
+      from = dateMath.parse(from, false, timeZone.raw as any);
```
Feels kind of weird having to cast this to `any`. Couldn't find any other solution since datemath defines `Timezone` in the following way:
https://github.com/grafana/grafana/blob/c0bb154603ec9a4bbc973e5b6b044918c7688f06/public/app/core/utils/datemath.ts#L6-L18

Is `type Timezone = 'utc';` correct?

hugohaggmark replied:
Maybe it should be `utc | browser`?

marefr replied:
Yes I had something similar in mind, but think that doesn't help since I don't have that constraint on my `TimeZone.raw` interface. Just noticed the comment of timezone for function

> * @param timezone Only string 'utc' is acceptable here, for anything else, local timezone is used. 

Seems like all other callers of this function which provides timezone is of type any so maybe wait with changing this for now?

torkelo replied:
feels so wrong that the Graph has to parse the time range, it never cares about the raw format. But not directly related to this PR, but part of the problem of passing RawTimeRange around in explore.

marefr replied:
Yes I'm currently trying to fix this

marefr committed:
add constant and factory method for TimeZone

davkal reviewed (commented):
Code looks great. 

Small issue:
The range string is still in local time when using UTC.
To repro, set UTC, run a query, then click on the left arrow next to the timepicker.

marefr commented:
Thanks @davkal will have a look at it.

marefr committed:
Merge remote-tracking branch 'origin/master' into 12812_explore_timezone

torkelo commented on public/app/features/explore/Graph.tsx:
```diff
@@ -79,6 +79,7 @@ interface GraphProps {
   width?: number;
   id?: string;
   range: RawTimeRange;
```
This component surely should use TimeRange and not have to parse it?

marefr replied:
Yes currently fixing this

marefr committed:
Merge remote-tracking branch 'origin/master' into 12812_explore_timezone

marefr committed:
Merge remote-tracking branch 'origin/master' into 12812_explore_timezone

marefr committed:
explore: time and time zone enhancements

Now uses TimeRange instead of RawTimeRange in explore item
state tree and only parsing actual time in a few action
handlers.
Time picker should now properly handle moving back/forward and
apply time range when both utc and non utc time zone.
URL range representation is changed from YYYY-MM-DD HH:mm:ss
to epoch ms.

marefr commented on public/app/features/explore/state/selectors.ts:
```diff
@@ -28,3 +31,43 @@ export const deduplicatedLogsSelector = createLodashMemoizedSelector(
+
+export const timeRangeFromUrlSelector = (range: RawTimeRange, timeZone: TimeZone): TimeRange => {
+  const raw = {
+    from: parseRawTime(range.from),
```
Need to do the parsing here since when going to Explore from dashboard `updateLocation` action is dispatched and it feld bad to do this parsing in the `updateLocation` reducer logic - should have no side effects right?

marefr commented on public/app/core/utils/explore.ts:
```diff
@@ -395,3 +385,11 @@ export const getQueryKeys = (queries: DataQuery[], datasourceInstance: DataSourc
   return queryKeys;
 };
+
+export const getTimeRange = (timeZone: TimeZone, rawRange: RawTimeRange): TimeRange => {
```
Should this be here or moved to selectors? I put it here since the function doesn't take a state object as argument like ExploreItemState, instead it takes the actual "properties" needed.

dprokop replied:
This is a legit place for such utils

marefr commented on public/app/features/explore/Explore.tsx:
```diff
@@ -111,12 +113,13 @@ export class Explore extends React.PureComponent<ExploreProps> {
-    const { exploreId, urlState, initialized } = this.props;
-    const { datasource, queries, range = DEFAULT_RANGE, ui = DEFAULT_UI_STATE } = (urlState || {}) as ExploreUrlState;
+    const { exploreId, urlState, initialized, timeZone } = this.props;
+    const { datasource, queries, range: urlRange, ui = DEFAULT_UI_STATE } = (urlState || {}) as ExploreUrlState;
```
Struggled a bit with this one. Seems to me it would have been more suitable if this logic was put in the initializeExplore action.  Then I could have just sent a DEFAULT_RANGE as initialRange and the action could have figured out if range should be taken from initialRange or urlstate. Thoughts?

I guess a general question here is where/how to handle default values since there's also DEFAULT_UI_STATE in use here in case of empty urlstate.

dprokop replied:
Agree that those default values could/should be handled on action level. I don't like directly accessing store in components like this:  `const initialDatasource = datasource || store.get(LAST_USED_DATASOURCE_KEY);` especially

marefr replied:
Thanks :+1: Let's solve this in another refactor. Will try and create issue to track this

marefr replied:
Just noticed you left a lot of other comments. GitHub doesn't send notifications from time to time :cry:

marefr replied:
Moved it down to `mapStateToProps`

marefr commented on public/app/features/explore/Graph.tsx:
```diff
@@ -133,35 +133,29 @@ export class Graph extends PureComponent<GraphProps, GraphState> {
       const range = {
-        from: moment(ranges.xaxis.from),
-        to: moment(ranges.xaxis.to),
+        from: timeZone.isUtc ? moment.utc(ranges.xaxis.from) : moment(ranges.xaxis.from),
```
Maybe Graph component should publish from/to in epoch ms instead of moment and the GraphContainer would be responsible for converting that to moment?

torkelo replied:
yes, that could make things easier I think

marefr commented on public/app/features/explore/Graph.tsx:
```diff
@@ -133,35 +133,29 @@ export class Graph extends PureComponent<GraphProps, GraphState> {
-    }
-    const min = from.valueOf();
-    const max = to.valueOf();
+    const min = range.from.valueOf();
```
Related to comment regarding `onPlotSelected` above, maybe this component should have numeric from/to props to represent epoch ms and GraphContainer would be responsible converting to epoch ms. This way we remove moment dependency from this component.

torkelo commented:
Looks good, like how time range is now parsed on init & change time actions

torkelo reviewed (approved):
Looks good. making onChangeTime on the graph be from/to in ms epoch sounds like a good idea

marefr commented on public/app/features/explore/TimePicker.test.tsx:
```diff
@@ -1,74 +1,238 @@
-    const wrapper = shallow(<TimePicker />);
-    expect(wrapper.find('.timepicker-rangestring').text()).toBe(rangeString);
-    expect(wrapper.find('.gf-timepicker-dropdown').exists()).toBe(false);
+  it('render default values when closed and relative time range', () => {
```
Not at all satisfied with these tests. Recommended to use snapshots instead to verify the rendered markup?

dprokop replied:
Since these are render specific tests I would opt for snapshots. Especially tat you rely on class names which 1) tend to change over time 2) sooner or later will be replaced with Emotion class names, which won't be that testable.

marefr commented on public/app/features/explore/state/actions.test.ts:
```diff
@@ -90,7 +109,10 @@ describe('refreshExplore', () => {
         expect(payload.eventBridge).toEqual(eventBridge);
         expect(payload.queries.length).toBe(1); // Queries have generated keys hard to expect on
-        expect(payload.range).toEqual(range);
+        expect(payload.range.from).toEqual(testRange.from);
```
What are we actually testing here?

marefr commented on public/app/features/explore/state/actions.ts:
```diff
@@ -235,12 +240,14 @@ export function initializeExplore(
 ): ThunkResult<void> {
-  return async dispatch => {
+  return async (dispatch, getState) => {
+    const timeZone = getTimeZone(getState().user);
```
Would it be a bad idea to dispatch `changeTime` action here instead?

marefr commented on public/app/features/explore/state/actions.ts:
```diff
@@ -235,12 +240,14 @@ export function initializeExplore(
-  return async dispatch => {
+  return async (dispatch, getState) => {
+    const timeZone = getTimeZone(getState().user);
+    const range = getTimeRange(timeZone, rawRange);
```
Would it be a bad idea to dispatch `changeTime` action here instead?

marefr commented on public/app/features/explore/state/actions.ts:
```diff
@@ -830,19 +854,20 @@ export function refreshExplore(exploreId: ExploreId): ThunkResult<void> {
       const initialQueries = ensureQueries(queries);
-      const initialRange = { from: parseTime(range.from), to: parseTime(range.to) };
-      dispatch(initializeExplore(exploreId, datasource, initialQueries, initialRange, containerWidth, eventBridge, ui));
+      dispatch(initializeExplore(exploreId, datasource, initialQueries, range, containerWidth, eventBridge, ui));
```
Related to comment in Explore.tsx, probably easier to let `initializeExplore` figure out what the actual range is and/or dispatch `changeTime` action.

marefr commented on public/app/features/explore/state/reducers.ts:
```diff
@@ -86,7 +86,11 @@ export const makeExploreItemState = (): ExploreItemState => ({
   queryTransactions: [],
   queryIntervals: { interval: '15s', intervalMs: DEFAULT_GRAPH_INTERVAL },
-  range: DEFAULT_RANGE,
+  range: {
```
Related to earlier comments regarding `initializeExplore` action I think we could use a null range as default state instead and by that hopefully moving default range to one single place.

davkal commented:
Issue I found. Repro: 

1. Setting user timezone to UTC, saving,
2. then back to Local Browser
3. then heading to Explore: UTC is in the timepicker.

jschill commented:
@davkal We're unable to reproduce that issue. Are you sure you saved after changing back to `Local browser`?

marefr committed:
explore: use AbsoluteTimeRange in graph component instead of moment

marefr commented:
@torkelo now I added a new interface to `@grafana/ui` called `AbsoluteTimeRange` which holds from/to as numbers. In the future I'm thinking that we can add an interface `RelativeTimeRange` which holds from/to as strings. These interfaces could then be passed around instead of today's RawTimeRange/TimeRange which includes moment.

torkelo commented:
Think that could be work, only problem will be dashboard time range and datasource plug-ins expecting a moment

dprokop commented on public/app/features/explore/state/selectors.ts:
```diff
@@ -28,3 +31,43 @@ export const deduplicatedLogsSelector = createLodashMemoizedSelector(
   }
 );
+
+const parseRawTime = (value): Moment | string => {
```
I think both parseRawTime and timeRangeFromUrlSelector does not belong to state/selectors.ts file. Selectors are basicaly functions that selects slices of redux state, while these two functions feel like utils to me.

marefr replied:
Fixed. Will push soon

dprokop commented on public/app/features/explore/Explore.tsx:
```diff
@@ -111,12 +113,13 @@ export class Explore extends React.PureComponent<ExploreProps> {
     const initialDatasource = datasource || store.get(LAST_USED_DATASOURCE_KEY);
     const initialQueries: DataQuery[] = ensureQueries(queries);
-    const initialRange = { from: parseTime(range.from), to: parseTime(range.to) };
+    const initialRange = urlRange ? timeRangeFromUrlSelector(urlRange, timeZone).raw : DEFAULT_RANGE;
```
I think you could do it on mapStateToProps level rather than here. Especially that you have all necessary information there already.

marefr replied:
It's done

marefr committed:
explore: fix time range passed to query row/editor

Also makes a copy of the time range passed to timeSrv to make sure immutability
of explore time range when for example elasticsearch test datasources uses
timeSrv and sets a time range of last 1 min

hugohaggmark commented on packages/grafana-ui/src/types/time.ts:
```diff
@@ -11,11 +11,29 @@ export interface TimeRange {
+  isUtc: boolean;
+}
+
+export const parseTimeZone = (raw: string): TimeZone => {
```
Maybe introduce a type for that?
```suggestion
type TimeZone = 'utc' | 'browser' | 'blabla';
export const parseTimeZone = (raw: TimeZone): TimeZone => {
```

marefr replied:
Decided to solve this later since it requires a lot of changes. Updated #16806

hugohaggmark commented on packages/grafana-ui/src/types/time.ts:
```diff
@@ -11,11 +11,29 @@ export interface TimeRange {
+
+export const parseTimeZone = (raw: string): TimeZone => {
+  return {
+    raw: raw,
```
Minor: the extra `raw` can be removed
```suggestion
    raw,
```

marefr replied:
Fixed. Will push soon

hugohaggmark commented on public/app/core/utils/explore.ts:
```diff
@@ -395,3 +385,11 @@ export const getQueryKeys = (queries: DataQuery[], datasourceInstance: DataSourc
+
+export const getTimeRange = (timeZone: TimeZone, rawRange: RawTimeRange): TimeRange => {
+  return {
+    from: dateMath.parse(rawRange.from, false, timeZone.raw as any),
```
If we extend `Timezone` type there is no need for `as any` here
```suggestion
    from: dateMath.parse(rawRange.from, false, timeZone.raw),
```

marefr replied:
Decided to solve this later since it requires a lot of changes. Updated #16806

hugohaggmark reviewed (approved):
LGTM! (again)

davkal commented:
Works now! Been changing wrong prefs. All systems 💚

marefr commented:
> Think that could be work, only problem will be dashboard time range and datasource plug-ins expecting a moment

@torkelo have added an epic to fix this long term #16806

marefr committed:
Merge remote-tracking branch 'origin/master' into 12812_explore_timezone

marefr committed:
explore: move timeRangeFromUrl from selectors to utils

marefr committed:
cleanup

marefr committed:
explore: move initializing code from componentDidMount to mapStateToProps
